File: options/bms_folder.py
import difflib
import os
import shutil
from typing import List, Optional, Tuple

from bms import BMSInfo, get_dir_bms_info
from fs import bms_dir_similarity
from fs.name import get_vaild_fs_name
from fs.move import REPLACE_OPTION_UPDATE_PACK, move_elements_across_dir


def append_artist_name_by_bms(root_dir: str):
    """该脚本适用于希望在作品文件夹名后添加“ [艺术家]”的情况。"""
    dir_names: List[str] = [
        dir_name
        for dir_name in os.listdir(root_dir)
        if os.path.isdir(os.path.join(root_dir, dir_name))
    ]

    pairs: List[Tuple[str, str]] = []

    for dir_name in dir_names:
        dir_path = os.path.join(root_dir, dir_name)
        if not os.path.isdir(dir_path):
            continue
        # Has been set?
        if dir_name.endswith("]"):
            continue
        bms_info: Optional[BMSInfo] = get_dir_bms_info(dir_path)
        if bms_info is None:
            print(f"Dir {dir_path} has no bms files!")
            continue
        new_dir_name = f"{dir_name} [{get_vaild_fs_name(bms_info.artist)}]"
        print("- Ready to rename: {} -> {}".format(dir_name, new_dir_name))
        pairs.append((dir_name, new_dir_name))

    selection = input("Do transfering? [y/N]:")
    if not selection.lower().startswith("y"):
        print("Aborted.")
        return

    for from_dir_name, target_dir_name in pairs:
        from_dir_path = os.path.join(root_dir, from_dir_name)
        target_dir_path = os.path.join(root_dir, target_dir_name)
        shutil.move(from_dir_path, target_dir_path)


def _workdir_append_name_by_bms(work_dir: str) -> bool:
    """
    该脚本适用于原有文件夹名与BMS文件无关内容的情况。
    会在文件夹名后添加“. 标题 [艺术家]”
    """
    if not os.path.split(work_dir)[-1].strip().isdigit():
        print(f"{work_dir} has been renamed! Skipping...")
        return False

    info: Optional[BMSInfo] = get_dir_bms_info(work_dir)
    if info is None:
        print(f"{work_dir} has no bms/bmson files!")
        return False

    # Deal with info
    print(f"{work_dir} found bms title: {info.title} artist: {info.artist}")
    title = info.title
    artist = info.artist

    # Rename
    new_dir_path = (
        f"{work_dir}. {get_vaild_fs_name(title)} [{get_vaild_fs_name(artist)}]"
    )
    shutil.move(work_dir, new_dir_path)
    return True


def append_name_by_bms(root_dir: str):
    """
    该脚本用于重命名作品文件夹。
    格式：“标题 [艺术家]”
    """
    fail_list = []
    for dir_name in os.listdir(root_dir):
        dir_path = os.path.join(root_dir, dir_name)
        if not os.path.isdir(dir_path):
            continue
        result = _workdir_append_name_by_bms(dir_path)
        if not result:
            fail_list.append(dir_name)
    if len(fail_list) > 0:
        print("Fail Count:", len(fail_list))
        print(fail_list)


def _set_workdir_name_by_bms(work_dir: str) -> bool:
    info: Optional[BMSInfo] = get_dir_bms_info(work_dir)
    while info is None:
        print(f"{work_dir} has no bms/bmson files! Trying to move out.")
        bms_dir_elements = os.listdir(work_dir)
        if len(bms_dir_elements) == 0:
            print(" - Empty dir! Deleting...")
            try:
                os.rmdir(work_dir)
            except PermissionError as e:
                print(e)
            return False
        if len(bms_dir_elements) != 1:
            print(f" - Element count: {len(bms_dir_elements)}")
            return False
        bms_dir_inner = os.path.join(work_dir, bms_dir_elements[0])
        if not os.path.isdir(bms_dir_inner):
            print(f" - Folder has only a file: {bms_dir_elements[0]}")
            return False
        print(" - Moving out files...")
        move_elements_across_dir(bms_dir_inner, work_dir)
        info = get_dir_bms_info(work_dir)

    parent_dir, _ = os.path.split(work_dir)
    if parent_dir is None:
        raise Exception("Parent is None!")

    if len(info.title) == 0 and len(info.artist) == 0:
        print(f"{work_dir}: Info title and artist is EMPTY!")
        return False

    # Rename
    new_dir_path = os.path.join(
        parent_dir,
        f"{get_vaild_fs_name(info.title)} [{get_vaild_fs_name(info.artist)}]",
    )

    # Same? Ignore
    if work_dir == new_dir_path:
        return True

    print(f"{work_dir}: Rename! Title: {info.title}; Artist: {info.artist}")
    if not os.path.isdir(new_dir_path):
        # Move Directly
        shutil.move(work_dir, new_dir_path)
        return True

    # Same dir?
    similarity = bms_dir_similarity(work_dir, new_dir_path)
    print(f" - Directory {new_dir_path} exists! Similarity: {similarity}")
    if similarity < 0.8:
        print(" - Merge canceled.")
        return False

    print(" - Merge start!")
    move_elements_across_dir(
        work_dir,
        new_dir_path,
        replace_options=REPLACE_OPTION_UPDATE_PACK,
    )
    return True


def set_name_by_bms(root_dir: str):
    """
    该脚本用于重命名作品文件夹。
    格式：“标题 [艺术家]”
    """
    fail_list = []
    for dir_name in os.listdir(root_dir):
        dir_path = os.path.join(root_dir, dir_name)
        if not os.path.isdir(dir_path):
            continue
        result = _set_workdir_name_by_bms(dir_path)
        if not result:
            fail_list.append(dir_name)
    if len(fail_list) > 0:
        print("Fail Count:", len(fail_list))
        print(fail_list)


def copy_numbered_workdir_names(root_dir_from: str, root_dir_to: str):
    """
    该脚本使用于以下情况：
    已经有一个文件夹A，它的子文件夹名为“”等带有编号+小数点的形式。
    现在有另一个文件夹B，它的子文件夹名都只有编号。
    将A中的子文件夹名，同步给B的对应的子文件夹。
    """
    src_dir_names = [
        dir_name
        for dir_name in os.listdir(root_dir_from)
        if os.path.isdir(os.path.join(root_dir_from, dir_name))
    ]
    # List Dst Dir
    for dir_name in os.listdir(root_dir_to):
        dir_path = os.path.join(root_dir_to, dir_name)
        # Get Num
        dir_num = dir_name.split(" ")[0].split(".")[0]
        if not dir_num.isdigit():
            continue
        # Search src name
        for src_name in src_dir_names:
            if not src_name.startswith(dir_num):
                continue
            # Rename
            target_dir_path = os.path.join(root_dir_to, src_name)
            print(f"Rename {dir_name} to {src_name}")
            shutil.move(dir_path, target_dir_path)
            break


def scan_folder_similar_folders(root_dir: str, similarity_trigger: float = 0.7):
    dir_names: List[str] = [
        dir_name
        for dir_name in os.listdir(root_dir)
        if os.path.isdir(os.path.join(root_dir, dir_name))
    ]
    print(f"当前目录下有{len(dir_names)}个文件夹。")
    # Sort
    dir_names.sort()
    # Scan
    for i, dir_name in enumerate(dir_names):
        if i == 0:
            continue
        former_dir_name = dir_names[i - 1]
        # 相似度
        similarity = difflib.SequenceMatcher(None, former_dir_name, dir_name).ratio()
        if similarity < similarity_trigger:
            continue
        print(f"发现相似项：{former_dir_name} <=> {dir_name}")


def undo_set_name(root_dir: str):
    for dir_name in os.listdir(root_dir):
        dir_path = os.path.join(root_dir, dir_name)
        if not os.path.isdir(dir_path):
            continue
        new_dir_name = dir_name.split(" ")[0]
        new_dir_path = os.path.join(root_dir, new_dir_name)
        if dir_name == new_dir_name:
            continue
        print(f"Rename {dir_name} to {new_dir_name}")
        shutil.move(dir_path, new_dir_path)

File: src/options/bms_folder.rs
//! BMS folder operations.
//!
//! This module provides functions for renaming and managing BMS work directories.

use std::path::Path;
use tracing::info;

use crate::bms::types::MEDIA_FILE_EXTS;

/// Append title and artist info to folder names based on BMS files.
///
/// This replicates Python's `append_name_by_bms(root_dir)`:
/// - Iterates through subdirectories
/// - Renames folders that are purely numeric to "num. title \[artist\]" format
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
pub async fn append_name_by_bms(root_dir: &Path) -> Result<(), std::io::Error> {
    if !root_dir.is_dir() {
        return Ok(());
    }

    let entries = std::fs::read_dir(root_dir)?;
    let mut to_rename: Vec<(PathBuf, PathBuf)> = Vec::new();

    for entry in entries.flatten() {
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }

        let dir_name = dir_path.file_name().and_then(|n| n.to_str()).unwrap_or("");

        // Skip if already renamed (has content after the number)
        if !dir_name.trim().is_empty()
            && dir_name
                .chars()
                .all(|c| c.is_ascii_digit() || ('\u{FF10}'..='\u{FF19}').contains(&c))
        {
            // This is a numeric-only folder, try to rename
            if let Some(new_name) = rename_folder_by_bms(&dir_path).await {
                let new_path = dir_path.with_file_name(&new_name);
                to_rename.push((dir_path, new_path));
            }
        }
    }

    for (from, to) in to_rename {
        info!("Renaming {:?} -> {:?}", from.file_name(), to.file_name());
        std::fs::rename(&from, &to)?;
    }

    Ok(())
}

/// Rename a single folder based on its BMS info
///
/// Returns the new folder name if renamed, `None` if skipped.
#[must_use]
async fn rename_folder_by_bms(work_dir: &Path) -> Option<String> {
    use crate::bms::dir::get_dir_bms_info;
    use crate::fs::name::get_valid_fs_name;

    let dir_name = work_dir.file_name().and_then(|n| n.to_str()).unwrap_or("");

    // Check if it's a purely numeric folder
    if !dir_name.trim().is_empty()
        && !dir_name
            .chars()
            .all(|c| c.is_ascii_digit() || ('\u{FF10}'..='\u{FF19}').contains(&c))
    {
        return None;
    }

    let info = get_dir_bms_info(work_dir).await?;
    if info.title.is_empty() && info.artist.is_empty() {
        return None;
    }

    let new_dir_name = format!(
        "{}. {} [{}]",
        dir_name.trim(),
        get_valid_fs_name(&info.title),
        get_valid_fs_name(&info.artist)
    );

    Some(new_dir_name)
}

/// Copy folder names from source to destination based on numeric prefix
///
/// This replicates Python's `copy_numbered_workdir_names(root_dir_from, root_dir_to)`:
/// - Source folders have format "num. title \[artist\]"
/// - Destination folders have format "num"
/// - Copies source names to destination based on numeric prefix match
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
pub fn copy_numbered_workdir_names(
    root_dir_from: &Path,
    root_dir_to: &Path,
) -> Result<(), std::io::Error> {
    if !root_dir_from.is_dir() || !root_dir_to.is_dir() {
        return Ok(());
    }

    // Get source directory names
    let src_names: Vec<(String, PathBuf)> = std::fs::read_dir(root_dir_from)?
        .filter_map(std::result::Result::ok)
        .filter(|e| e.path().is_dir())
        .filter_map(|e| e.file_name().to_str().map(|n| (n.to_string(), e.path())))
        .collect();

    // Iterate destination directories
    for entry in std::fs::read_dir(root_dir_to)?.flatten() {
        let dst_path = entry.path();
        if !dst_path.is_dir() {
            continue;
        }

        let dst_name = dst_path.file_name().and_then(|n| n.to_str()).unwrap_or("");

        // Extract numeric prefix
        let num_prefix = dst_name.split_whitespace().next().unwrap_or("");
        let numeric_part = num_prefix.split('.').next().unwrap_or("");

        if !numeric_part.chars().all(|c| c.is_ascii_digit()) {
            continue;
        }

        // Search for matching source folder
        for (src_name, _src_path) in &src_names {
            if src_name.starts_with(numeric_part) {
                let target_path = dst_path.with_file_name(src_name);
                if target_path != dst_path {
                    info!(
                        "Renaming {:?} -> {:?}",
                        dst_path.file_name(),
                        target_path.file_name()
                    );
                    std::fs::rename(&dst_path, &target_path)?;
                }
                break;
            }
        }
    }

    Ok(())
}

/// Append artist name to folder names based on BMS files
///
/// This replicates Python's `append_artist_name_by_bms(root_dir)`:
/// - Adds " \[artist\]" suffix to folders not already ending with "\]"
/// - Shows confirmation before renaming
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
///
/// # Panics
///
/// Panics if stdout flush or stdin read fails.
pub async fn append_artist_name_by_bms(root_dir: &Path) -> Result<(), std::io::Error> {
    use std::io::{self, Write};

    use crate::bms::dir::get_dir_bms_info;
    use crate::fs::name::get_valid_fs_name;

    if !root_dir.is_dir() {
        return Ok(());
    }

    let entries: Vec<_> = std::fs::read_dir(root_dir)?
        .filter_map(std::result::Result::ok)
        .collect();

    let mut pairs: Vec<(PathBuf, PathBuf)> = Vec::new();

    for entry in entries {
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }

        let dir_name = dir_path.file_name().and_then(|n| n.to_str()).unwrap_or("");

        // Skip if already has artist suffix
        if dir_name.ends_with(']') {
            continue;
        }

        let info = get_dir_bms_info(&dir_path).await;
        if info.is_none() {
            println!("Dir {} has no bms files!", dir_path.display());
            continue;
        }

        let info = info.unwrap();
        let new_dir_name = format!("{dir_name} [{}]", get_valid_fs_name(&info.artist));
        println!("- Ready to rename: {dir_name} -> {new_dir_name}");
        pairs.push((dir_path, root_dir.join(&new_dir_name)));
    }

    if pairs.is_empty() {
        println!("No folders to rename");
        return Ok(());
    }

    print!("Do transferring? [y/N]: ");
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    if !input.trim().to_lowercase().starts_with('y') {
        println!("Aborted.");
        return Ok(());
    }

    for (from, to) in pairs {
        std::fs::rename(&from, &to)?;
    }

    Ok(())
}

/// Set folder names based on BMS info (title \[artist\] format)
///
/// This replicates Python's `set_name_by_bms(root_dir)`:
/// - Renames folders to "title \[artist\]" format
/// - Handles merging if target already exists (with similarity check)
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
pub async fn set_name_by_bms(root_dir: &Path) -> Result<(), std::io::Error> {
    if !root_dir.is_dir() {
        return Ok(());
    }

    let mut fail_list: Vec<String> = Vec::new();

    let entries: Vec<_> = std::fs::read_dir(root_dir)?
        .filter_map(std::result::Result::ok)
        .collect();

    for entry in entries {
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }

        let dir_name = dir_path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("")
            .to_string();

        if !set_single_folder_name_by_bms(&dir_path).await? {
            fail_list.push(dir_name);
        }
    }

    if !fail_list.is_empty() {
        println!("Fail Count: {}", fail_list.len());
        for name in &fail_list {
            println!("  {name}");
        }
    }

    Ok(())
}

/// Set a single folder's name based on its BMS info
/// Returns true if successful, false if skipped or failed
async fn set_single_folder_name_by_bms(work_dir: &Path) -> Result<bool, std::io::Error> {
    use crate::bms::dir::get_dir_bms_info;
    use crate::fs::name::{bms_dir_similarity, get_valid_fs_name};
    use crate::fs::pack_move::{
        MoveOptions, REPLACE_OPTION_UPDATE_PACK, ReplaceOptions, move_elements_across_dir,
    };

    let mut info = get_dir_bms_info(work_dir).await;

    // If no BMS info found, try to move out nested contents and retry
    while info.is_none() {
        println!(
            "{} has no bms/bmson files! Trying to move out.",
            work_dir.display()
        );

        let elements: Vec<_> = std::fs::read_dir(work_dir)?
            .filter_map(std::result::Result::ok)
            .collect();

        if elements.is_empty() {
            println!(" - Empty dir! Deleting...");
            match std::fs::remove_dir(work_dir) {
                Ok(()) => {}
                Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {
                    println!(" x PermissionError: {e}");
                }
                Err(e) => return Err(e),
            }
            return Ok(false);
        }

        if elements.len() != 1 {
            println!(" - Element count: {}", elements.len());
            return Ok(false);
        }

        let inner_path = &elements[0].path();
        if !inner_path.is_dir() {
            println!(" - Folder has only a file: {:?}", inner_path.file_name());
            return Ok(false);
        }

        println!(" - Moving out files...");
        move_elements_across_dir(
            inner_path,
            work_dir,
            MoveOptions::default(),
            &ReplaceOptions::default(),
        )?;
        info = get_dir_bms_info(work_dir).await;
    }

    let info = info.unwrap();
    let parent_dir = work_dir.parent().unwrap_or(work_dir);

    if info.title.is_empty() && info.artist.is_empty() {
        println!("{}: Info title and artist is EMPTY!", work_dir.display());
        return Ok(false);
    }

    let new_dir_path = parent_dir.join(format!(
        "{} [{}]",
        get_valid_fs_name(&info.title),
        get_valid_fs_name(&info.artist)
    ));

    // Same directory?
    if work_dir == new_dir_path {
        return Ok(true);
    }

    println!(
        "{}: Rename! Title: {}; Artist: {}",
        work_dir.display(),
        info.title,
        info.artist
    );

    if !new_dir_path.is_dir() {
        std::fs::rename(work_dir, &new_dir_path)?;
        return Ok(true);
    }

    // Directory already exists - check similarity
    let similarity = bms_dir_similarity(work_dir, &new_dir_path);
    println!(
        " - Directory {} exists! Similarity: {similarity}",
        new_dir_path.display()
    );

    if similarity < 0.8 {
        println!(" - Merge canceled.");
        return Ok(false);
    }

    println!(" - Merge start!");
    move_elements_across_dir(
        work_dir,
        &new_dir_path,
        MoveOptions::default(),
        &REPLACE_OPTION_UPDATE_PACK,
    )?;
    Ok(true)
}

/// Scan for similar folder names
///
/// This replicates Python's `scan_folder_similar_folders(root_dir, similarity_trigger)`:
/// - Uses difflib.SequenceMatcher to find similar folder names
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
pub fn scan_folder_similar_folders(
    root_dir: &Path,
    similarity_trigger: f64,
) -> Result<(), std::io::Error> {
    if !root_dir.is_dir() {
        return Ok(());
    }

    let dir_names: Vec<String> = std::fs::read_dir(root_dir)?
        .filter_map(std::result::Result::ok)
        .filter(|e| e.path().is_dir())
        .filter_map(|e| e.file_name().to_str().map(String::from))
        .collect();

    println!("当前目录下有{}个文件夹。", dir_names.len());

    let mut sorted_names = dir_names.clone();
    sorted_names.sort();

    for i in 1..sorted_names.len() {
        let former = &sorted_names[i - 1];
        let current = &sorted_names[i];

        let similarity = similar_ratio(former, current);
        if similarity < similarity_trigger {
            continue;
        }
        println!("发现相似项：{former} <=> {current}");
    }

    Ok(())
}

/// Calculate similarity ratio between two strings.
/// Returns a value between 0.0 and 1.0.
fn similar_ratio(a: &str, b: &str) -> f64 {
    sequence_matcher_ratio(a, b)
}

fn sequence_matcher_ratio(a: &str, b: &str) -> f64 {
    let a_chars: Vec<char> = a.chars().collect();
    let b_chars: Vec<char> = b.chars().collect();
    if a_chars.is_empty() && b_chars.is_empty() {
        return 1.0;
    }
    if a_chars.is_empty() || b_chars.is_empty() {
        return 0.0;
    }
    let total = a_chars.len() + b_chars.len();
    let matches = find_longest_match(&a_chars, &b_chars);
    #[expect(clippy::cast_precision_loss)]
    {
        (2 * matches) as f64 / total as f64
    }
}

#[allow(clippy::similar_names)]
fn find_longest_match(a: &[char], b: &[char]) -> usize {
    let mut best_len = 0;
    let mut best_ai = 0;
    let mut best_bi = 0;

    for ai in 0..a.len() {
        for bi in 0..b.len() {
            let mut k = 0;
            while ai + k < a.len() && bi + k < b.len() && a[ai + k] == b[bi + k] {
                k += 1;
            }
            if k > best_len {
                best_len = k;
                best_ai = ai;
                best_bi = bi;
            }
        }
    }

    if best_len == 0 {
        return 0;
    }

    let left_matches = find_longest_match(&a[..best_ai], &b[..best_bi]);
    let right_matches = find_longest_match(&a[best_ai + best_len..], &b[best_bi + best_len..]);

    best_len + left_matches + right_matches
}

/// Undo `set_name` by removing " \[artist\]" suffix
///
/// This replicates Python's `undo_set_name(root_dir)`:
/// - Removes " \[artist\]" part from folder names
/// - Restores the original numeric prefix
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
pub fn undo_set_name(root_dir: &Path) -> Result<(), std::io::Error> {
    if !root_dir.is_dir() {
        return Ok(());
    }

    let entries: Vec<_> = std::fs::read_dir(root_dir)?
        .filter_map(std::result::Result::ok)
        .collect();

    for entry in entries {
        let dir_path = entry.path();
        if !dir_path.is_dir() {
            continue;
        }

        let dir_name = dir_path.file_name().and_then(|n| n.to_str()).unwrap_or("");

        let parts: Vec<&str> = dir_name.splitn(2, ' ').collect();
        let new_dir_name = parts[0];

        if dir_name == new_dir_name {
            continue;
        }

        let new_dir_path = root_dir.join(new_dir_name);

        if new_dir_path.is_dir() {
            println!(
                "Warning: Target {} already exists! Skipping {dir_name}",
                new_dir_path.display()
            );
            continue;
        }

        println!("Rename {dir_name} to {new_dir_name}");
        std::fs::rename(&dir_path, &new_dir_path)?;
    }

    Ok(())
}

/// Remove zero-sized media files and temp files
///
/// This replicates Python's `remove_zero_sized_media_files(current_dir, print_dir)`:
///
/// # Errors
///
/// Returns [`std::io::Error`] if directory operations fail.
pub fn remove_zero_sized_media_files(
    current_dir: &Path,
    print_dir: bool,
) -> Result<(), std::io::Error> {
    if print_dir {
        println!("Entering dir: {}", current_dir.display());
    }

    if !current_dir.is_dir() {
        println!("Not a vaild dir! Aborting...");
        return Ok(());
    }

    let mut next_dirs: Vec<PathBuf> = Vec::new();

    let entries: Vec<_> = std::fs::read_dir(current_dir)?
        .filter_map(std::result::Result::ok)
        .collect();

    for entry in &entries {
        let element_path = entry.path();
        let element_name = element_path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("");

        if element_path.is_file() {
            let is_temp_file = element_name.to_lowercase() == "desktop.ini"
                || element_name.to_lowercase() == "thumbs.db"
                || element_name.to_lowercase() == ".ds_store"
                || element_name.starts_with(".trash-")
                || element_name.starts_with("._");

            if is_temp_file {
                match std::fs::remove_file(&element_path) {
                    Ok(()) => println!(" - Remove temp file: {}", element_path.display()),
                    Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {
                        println!(" x PermissionError!");
                    }
                    Err(_) => {}
                }
                continue;
            }

            if !MEDIA_FILE_EXTS
                .iter()
                .any(|ext| element_name.ends_with(*ext))
            {
                continue;
            }

            match element_path.metadata() {
                Ok(metadata) if metadata.len() == 0 => match std::fs::remove_file(&element_path) {
                    Ok(()) => {
                        println!(" - Remove empty file: {}", element_path.display());
                    }
                    Err(e) if e.kind() == std::io::ErrorKind::PermissionDenied => {
                        println!(" x PermissionError!");
                    }
                    Err(_) => {}
                },
                _ => {}
            }
        } else if element_path.is_dir() {
            next_dirs.push(element_path);
        }
    }

    for next_dir in next_dirs {
        remove_zero_sized_media_files(&next_dir, print_dir)?;
    }

    Ok(())
}

use std::path::PathBuf;
